Issue #73: New app detection, tagging and category

bobheadxi opened the pull request:
- On initial AppInfo load, parses past 4 revisions and tags the newly added apps as "NEW"
- Added a new primary category under "All" that lists the "New" apps
- Other tweaks
for https://github.com/d4rken/reddit-android-appstore/issues/73

d4rken commented on app/src/main/java/subreddit/android/appstore/screens/navigation/CategoryFilter.java:
```diff
@@ -99,9 +112,10 @@ public int hashCode() {
         else if (secondaryCategory != null) return secondaryCategory;
         else if (primaryCategory != null) return primaryCategory;
-        else return context.getString(R.string.app_category_everything);
+        else return "All";
```
Why hardcode this and remove the string reference?

d4rken commented on app/src/main/java/subreddit/android/appstore/backend/reddit/wiki/LiveWikiRepository.java:
```diff
@@ -73,25 +77,103 @@ public void refresh() {
+                                                         @Query("page") String page,
+                                                         @Query("v") String id);
+
+        // Idea: could use this instead? not sure what it does though
```
It returns the difference between two revision ids. This could be used instead, to save traffic.
So instead of getting the whole wiki revision and then seeing when there is a app name, you could only get what was changed and see if a new app name is contained the changes. Not critical though it works how it is then we can use that and someone can improve upon this with a new PR.

bobheadxi replied:
Yeah I was just thinking about that as I finished up - my current implementation seems to add a decent amount of load time. I'll give it a try over the weekend, and remove the hard-coded string from your other comment when I'm done, forgot to fix that. Thanks, please let me know if there's anything else I can improve on!

_Sent from my Sony Xperia Z3 using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

d4rken commented:
Looks good otherwise, nice work!

Switching to the other call that just gets the difference would make it pretty fast I think. Just check the diff for new apps, check if the new apps are already known if not, tag them.

I didn't check code formatting explicitly, you ran the default Android Studio formatter on it, right?

bobheadxi commented:
@d4rken Thanks, I'll see what I can do. And yep, just used the the default Android Studio formatter

_Sent from my Sony Xperia Z3 using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

bobheadxi commented:
@d4rken I fixed the string references and added translations as well. On the subject of using the get difference API, however, I still can't work out what it does - using www.hurl.it, I tested the request https://oauth.reddit.com/r/Android/wiki/page&v&v2?v=bc0dd3d8-1a81-11e7-bff6-0aed84ec8996&v2=200d24f0-1a7e-11e7-b3e4-121bb5573bcc&page=apps which should get the difference between those two revisions. The response, however, just seems to be the entire wiki, and I can't seem to find any further help or documentation online.

d4rken commented on app/src/main/java/subreddit/android/appstore/backend/reddit/wiki/parser/SimpleBodyParser.java:
```diff
@@ -29,15 +28,16 @@ public SimpleBodyParser(EncodingFixer encodingFixer) {
     }
 
-    private static final Pattern CATEGORY_PATTERN = Pattern.compile("^(#+)(?:\\s*+)(.+?)$");
+    //private static final Pattern CATEGORY_PATTERN = Pattern.compile("^(#+)(?:\\s*+)(.+?)$");
```
Why leave in outcommented code?

d4rken commented:
I currently don't have the time to look into the reddit api. You could ask on reddit. Maybe someone else can help? 

How much longer is the load time now?

bobheadxi commented:
With the method in the previous commit, the tagging of new apps took about 18500ms with getWikiRevision and 14500ms with getWikiRevisionDiff (still have no idea why - I posted on r/redditdev but I doubt I'll get a response)

However, not sure why I didn't think of this before, but with this commit I changed the tagging so that it simply fetches the oldest revision requested (as defined by NUMOFREVISIONS, currently 6), then does a String.contains search for each app name in infos. This brought the additional load time required to tag the new apps down significantly to 710ms.

d4rken commented:
> This brought the additional load time required to tag the new apps down significantly to 710ms.

Nice,

d4rken commented:
Would you mind looking at #74 ? And then we release v0.7.0?

bobheadxi commented:
Will do!

_Sent from my Sony Xperia Z3 using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

d4rken commented:
@bobheadxi Did you ever find out what was going wrong with the diff API?

https://www.reddit.com/r/redditdev/comments/6ihoy9/how_does_get_wikipage_revision_difference_work/

bobheadxi commented:
No, I was thinking about finding out where to report it, but I never got around to it, since the current parsing is pretty quick already. I'm still not certain it's a bug though - I wonder how the web page version ie https://www.reddit.com/r/Android/wiki/apps?v=aee0d142-5b77-11e7-8310-0a32db770d50&v2=4587db92-27d2-11e7-be64-121ac739a4a0 works, not sure if there's any way to look into it
